zrml/prediction-markets/src/lib.rs

```diff
@@ -1824,6 +1926,35 @@ mod pallet {
             Ok(())
         }
 
+        fn market_close_manager<F>(now: T::BlockNumber, mut mutation: F) -> DispatchResult
+        where
+            F: FnMut(
+                &MarketIdOf<T>,
+                Market<T::AccountId, T::BlockNumber, MomentOf<T>>,
+            ) -> DispatchResult,
+        {
+            for market_id in MarketIdsPerCloseBlock::<T>::get(&now).iter() {
+                let market = T::MarketCommons::market(market_id)?;
+                mutation(market_id, market)?;
+            }
+            MarketIdsPerCloseBlock::<T>::remove(&now);
+
+            let current_time_frame = Self::calculate_time_frame_of_moment(T::MarketCommons::now());
+            // On first pass, we use current_time - 1 to ensure that the chain doesn't try to check
+            // all time frames since epoch.
+            let last_time_frame =
+                LastTimeFrame::<T>::get().unwrap_or_else(|| current_time_frame.saturating_sub(1));
+            for time_frame in last_time_frame.saturating_add(1)..=current_time_frame {
+                for market_id in MarketIdsPerCloseTimeFrame::<T>::get(&time_frame).iter() {
+                    let market = T::MarketCommons::market(market_id)?;
+                    mutation(market_id, market)?;
+                }
+                MarketIdsPerCloseTimeFrame::<T>::remove(&time_frame);
+            }
+            LastTimeFrame::<T>::set(Some(current_time_frame));
+            Ok(())
+        }
+
```

Comment:
I think there might be a problem when adding 1 to the `last_time_frame` in the extreme case of markets that run 1 block (which obviously don't make any sense, but I just want to point that out since it can lead to storage bloating and locked funds):
Assume we have the timestamp `12000` (for the sake of simplicity). The `on_initialize` hook runs and stores `LastTimeFrame = 1`. Now the next block is published, the block was created a bit faster than expected (can happen), so the timestamp is `23590`. It will divide `23590 / 12000 = 1`, since integer divisions are floored by default. During `on_initialize`, it will use the range `2..1`, since the `LastTimeFrame` was `1` and is incremented by `1` in the range, effectively skipping all markets that were created in the last block and that should end in the current block.

Reply:
Basically the problem is that there can be multiple blocks per time frame, right? What's the worst-case scenario, i.e. how many blocks per time frame could happen? Can we assume that each block will take at least 6sec to produce? Then it's two blocks per time frame, right?

Reply:
Yes, the essential problem is that there can be multiple blocks per time frame. I cannot answer definitely how many there are, there is also some allowed discrepancy between the timestamp that the current block author provides and that other block authors have locally. I have only seen two blocks in a time frame so far though. In any case, we should not write code that is based on such assumptions (as of how many blocks can fit into one frame).

I still think the problem persists. Here is a more elaborate example:

1. Block 0: Timestamp is `0`, we create a market *FIRSTMARKET* that ends on `12000`, `MarketIdsPerCloseTimeFrame(1)` is appended by *FIRSTMARKET*
2. Block 1: Timestamp is `12000`, `on_initialize` clears all markets in time frame `1..=1` including *FIRSTMARKET*, sets `LastTimeFrame` to `1`
3. Block 1: Timestamp is `12000`, user creates market *SECONDMARKET* that ends on `20000`, `MarketIdsPerCloseTimeFrame(1)` is appended by *SECONDMARKET*
4. Block 2: Timestamp is `23590`, `on_initialize` clears all markets in time frame `2..=1`, *SECONDMARKET* was never handled.

Reply:
Thanks Malte, all concerns are lifted now, well done!